crawler/helpers/cherrypicker_test_data.py

```diff
@@ -48,10 +53,33 @@ def flatten(nested_list: list) -> list:
 
 def generate_baracoda_barcodes(config: Config, num_required: int) -> list:
     baracoda_url = f"{config.BARACODA_BASE_URL}/barcodes_group/{BARACODA_PREFIX}/new?count={num_required}"
-    response = requests.post(baracoda_url)
-    response_json = response.json()
-    barcodes: list = response_json["barcodes_group"]["barcodes"]
-    return barcodes
+
+    retries = config.BARACODA_RETRY_ATTEMPTS
+    except_obj = None
+    response_json = None
+    while retries > 0:
+        try:
+            response = requests.post(baracoda_url)
+            if response.status_code == HTTPStatus.CREATED:
+                response_json = response.json()
+                barcodes: list = response_json["barcodes_group"]["barcodes"]
+                return barcodes
+            else:
+                retries = retries - 1
+                logger.error(TEST_DATA_ERROR_BARACODA_COG_BARCODES)
+                logger.error(response.json())
+                except_obj = CherrypickerDataError(TEST_DATA_ERROR_BARACODA_COG_BARCODES)
+        except requests.ConnectionError as e:
+            retries = retries - 1
+            logger.error(TEST_DATA_ERROR_BARACODA_CONNECTION)
+            except_obj = CherrypickerDataError(f"{TEST_DATA_ERROR_BARACODA_CONNECTION} -- {str(e)}")
+        except Exception:
+            retries = retries - 1
+            logger.error(TEST_DATA_ERROR_BARACODA_UNKNOWN)
+
+    if except_obj is not None:
+        raise except_obj
```

Comment:
If the second try hits a connection error, and the third-and-final try hits a generic exception, won't this raise the out-of-date connection error, because `except_obj` wasn't reset?

Reply:
Yes I expect so.  I didn't write this so better to ask @emrojo and @Sangeetha-Bheeman .  I just extracted the hard-coded strings.  It may be that the behaviour you described is what we want.  CherrypickerDataErrors are treated a bit more usefully for the user than other types of error.  The message is deemed safe to show the user.